### ISSUE TITLE
[FEATURE] Harvester node driver for rke and rancher 2.6

### ISSUE THREAD

**Issue by guangbochen** [harvester/harvester#861]
**Is your feature request related to a problem? Please describe.**
Refactor the Harvester node driver to support Rancher 2.6 and spin up the k8s cluster with rke2, additionally uses vanilla k8s API instead of the steve API for better HCI 2.0 standardization.

**Describe the solution you'd like**
- [x] Submit the Harvester node driver UI as built-in to the Rancher 2.6 dashboard. 
       - [ ] RKE1 https://github.com/rancher/dashboard/issues/3374 - as GA
       - [x] RKE2 https://github.com/rancher/dashboard/issues/3375 - as experimental
- [x] Add the default Harvester node driver to the Rancher. https://github.com/rancher/rancher/pull/33531
- [x] Support to spin up a Kubernetes cluster in the Harvester with RKE1 & RKE2 provisioned. [harvester-node-driver v0.2.0](https://github.com/harvester/docker-machine-driver-harvester/releases/tag/v0.2.0)
- [x] Support both imported and external Harvester clusters using the Kubeconfig file. [harvester-node-driver v0.2.0](https://github.com/harvester/docker-machine-driver-harvester/releases/tag/v0.2.0)

**Describe alternatives you've considered**
N/A

**Additional context**
https://github.com/rancher/dashboard/issues/3375

**Comment by belgaied2** [comment]
It might be a good idea to re-use [my docker-machine driver](https://github.com/belgaied2/docker-machine-driver-harvester) that I created for Harvester v0.1.0 based on the client-go for Kubernetes!

**Comment by futuretea** [comment]
# RKE1

When Rancher calls Node Driver to create clusters, a jail directory is created for each cluster.
```bash
root@c5d48a1006e2:/opt/jail# ls
c-dd4zk  c-kfp75  c-mpnfs  c-nmfs8  driver-jail
root@c5d48a1006e2:/opt/jail/c-nmfs8# ls
bin  dev  done  etc  kustomize.sh  lib  lib64  management-state  tmp  usr  var
```
After that, Rancher will chroot to the jail directory and create the nodes of the cluster. 

details:
https://github.com/rancher/rancher/blob/master/package/jailer.sh
https://github.com/rancher/rancher/blob/master/pkg/jailer/jailer.go
https://github.com/rancher/rancher/blob/master/pkg/jailer/jailer_linux.go
 https://github.com/rancher/rancher/blob/master/pkg/controllers/management/node/controller.go#L178

Because of this chroot mechanism, Node Driver will not be able to get InClusterConfig even if Rancher and Harvester are in the same cluster.

According to the [code of pod-driver](https://github.com/rancher/machine/blob/master/drivers/pod/pod.go), it should be a kind of driver that uses pod as node, which also has the above problem.

Therefore, the current docker-machine-driver-harvester implementation （PR harvester/docker-machine-driver-harvester#4） requires the user to pass the contents of the kubeconfig file to the Node Driver when calling the Node Driver.

This is required for both internal and external clusters.

**Comment by futuretea** [comment]
# RKE2
Node Driver uses the Kubernetes API instead of the Steve API, so Kubeconfig is now required to access the Harvester cluster.

<details>

<summary>If the Harvester cluster has already been imported by Rancher, then the user only needs to select the cluster and the filling of Kubeconfig is done automatically by the UI.</summary>

![image](https://user-images.githubusercontent.com/15064560/122802345-d78c8e00-d2f7-11eb-9000-5ca3efccbf1c.png)

</details>
<details>
<summary>And on Rancher's UI, users can easily select Harvester custom resources through drop-down boxes, such as images, networks, and namespaces.</summary>

![image](https://user-images.githubusercontent.com/15064560/122802417-ebd08b00-d2f7-11eb-855c-2e98a8043c99.png)

</details>
<details>
<summary>For Harvester clusters that are not managed by Rancher, users need to manually paste Kubeconfig, fill in the image name, network name, vm namespace and so on.</summary>

![image](https://user-images.githubusercontent.com/15064560/122802377-e07d5f80-d2f7-11eb-8b2d-92d0da4fe8ce.png)

</details>

**Comment by guangbochen** [comment]
**To Test:**
- launch a new rancher server using the `master-head` image https://rancher.com/docs/rancher/v2.x/en/quick-start-guide/deployment/quickstart-manual-setup/#2-install-rancher
- importing the downstream Harvester cluster(use master-head ISO) to the rancher server through the rancher dashboard UI
    - Click`Import Existing` and choose Generic
    - fill in an arbitrary cluster name and click Create
    - wait a few seconds(and refresh the page), you will see a Registration tab on the cluster detail page
    - copy the registration command(The second curl --insecure ...|kubectl apply, basically it is for deploying the agent to the downstream harvester cluster), run it against the harvester cluster
    - wait a few seconds and the harvester cluster should be imported and managed by the external Rancher.
- Go to the global settings via the hamburger menu and replace the rancher `ui-index` and `ui-dashboard-index` respectively (Note: this is not required for the official Rancher 2.6 release, currently the dashboard UI PR is not merged yet)
    - ui-index: https://rancher-harvester.s3.ap-southeast-1.amazonaws.com/harvester-node-driver/latest2/index.html
    - ui-dashboard-index:  https://rancher-harvester.s3.ap-southeast-1.amazonaws.com/harvester-node-driver/harvester-node-driver/index.html
 - verify the user is able to spin up the guest k8s cluster with Harvester node driver with both RKE1 and RKE2 provisioned(RKE1 as GAed / RKE2 as experimental (this depends on Rancher RKE2 GA capability).
    - select create `Harvester cluster` and add a cloud credential by choosing one of the imported Harvester clusters.
![image](https://user-images.githubusercontent.com/4569037/125890367-914a97ac-9bd7-4545-a67e-1b1a1ec5807c.png)
    - config the VM instance options(only cloud image is supported, and VLAN network is required), u specify the k8s config or you can add more node-pools if needed.
 ![image](https://user-images.githubusercontent.com/4569037/125901317-56a64316-8d67-47de-a23d-cf9016e4dd71.png)
    - Click create to spin up the guest k8s cluster and wait for the cluster to be ready.

**Comment by juadk** [comment]
1. Harvester cluster imported successfully in a master-head rancher: ![image](https://user-images.githubusercontent.com/6025636/126183356-42af0fe4-5eb7-4229-8a51-b64eca8f1575.png)

2. Updated the `ui-index` and `ui-dashboard-index`

3. The harvester node driver was inactive by default (expected?)
* spin up with RKE1: It fails, new issue reported [here](https://github.com/harvester/harvester/issues/1030)
* spin up with RKE2: It fails, new issue reported [here](https://github.com/harvester/harvester/issues/1031)

**Comment by futuretea** [comment]
> The harvester node driver was inactive by default (expected?)

If a driver need to be downloaded from Intenet, set it to be inactive by default should be better.

**Comment by juadk** [comment]
**Test to spin up a cluster with RKE2: FAILED**

![image](https://user-images.githubusercontent.com/6025636/126325404-110152b9-1610-4943-abd6-19e63a602344.png)
2 nodes keep pending on "Waiting on probes: calico"

However, all looks good from the cluster (named rke2) view:
![image](https://user-images.githubusercontent.com/6025636/126326178-3cbab325-2257-4060-8105-2c8f600e531d.png)

I opened a kubectl shell and looked at the pods status, all looks good too:
```
> kubectl get pods -A
NAMESPACE             NAME                                                              READY   STATUS      RESTARTS   AGE
calico-system         calico-kube-controllers-5f8b57fc77-742vd                          1/1     Running     0          82m
calico-system         calico-node-2hkq6                                                 1/1     Running     0          82m
calico-system         calico-node-l7bwz                                                 1/1     Running     0          78m
calico-system         calico-node-v25gw                                                 1/1     Running     0          76m
calico-system         calico-typha-66b479cdf8-b57mj                                     1/1     Running     0          82m
calico-system         calico-typha-66b479cdf8-nt9xd                                     1/1     Running     1          76m
calico-system         calico-typha-66b479cdf8-xdnkj                                     1/1     Running     1          78m
cattle-fleet-system   fleet-agent-84ccfd47b-6mjn5                                       1/1     Running     0          79m
cattle-system         apply-system-agent-upgrader-on-rke2-pool1-gw6nk-with-3b7f-l5wzm   1/1     Running     0          76m
cattle-system         apply-system-agent-upgrader-on-rke2-pool1-nkjbq-with-3b7f-hg2q8   1/1     Running     0          78m
cattle-system         apply-system-agent-upgrader-on-rke2-pool1-qmtm2-with-3b7f-nrb8b   1/1     Running     0          78m
cattle-system         cattle-cluster-agent-55bbb75957-c92cz                             1/1     Running     0          77m
cattle-system         dashboard-shell-zz8sz                                             2/2     Running     0          48s
cattle-system         system-upgrade-controller-8bbcd68fb-24w2l                         1/1     Running     0          79m
kube-system           etcd-rke2-pool1-gw6nk                                             1/1     Running     0          76m
kube-system           etcd-rke2-pool1-nkjbq                                             1/1     Running     0          77m
kube-system           etcd-rke2-pool1-qmtm2                                             1/1     Running     0          84m
kube-system           helm-install-rke2-calico-crd-fvpw5                                0/1     Completed   0          84m
kube-system           helm-install-rke2-calico-kbf7h                                    0/1     Completed   1          84m
kube-system           helm-install-rke2-coredns-jlcp9                                   0/1     Completed   0          84m
kube-system           helm-install-rke2-ingress-nginx-z5mn8                             0/1     Completed   0          84m
kube-system           helm-install-rke2-kube-proxy-vzcvs                                0/1     Completed   0          84m
kube-system           helm-install-rke2-metrics-server-q9l46                            0/1     Completed   0          84m
kube-system           kube-apiserver-rke2-pool1-gw6nk                                   1/1     Running     0          76m
kube-system           kube-apiserver-rke2-pool1-nkjbq                                   1/1     Running     0          76m
kube-system           kube-apiserver-rke2-pool1-qmtm2                                   1/1     Running     0          84m
kube-system           kube-controller-manager-rke2-pool1-gw6nk                          1/1     Running     0          76m
kube-system           kube-controller-manager-rke2-pool1-nkjbq                          1/1     Running     0          77m
kube-system           kube-controller-manager-rke2-pool1-qmtm2                          1/1     Running     0          84m
kube-system           kube-proxy-r27b6                                                  1/1     Running     0          78m
kube-system           kube-proxy-tx5fd                                                  1/1     Running     0          76m
kube-system           kube-proxy-vxqff                                                  1/1     Running     0          83m
kube-system           kube-scheduler-rke2-pool1-gw6nk                                   1/1     Running     0          76m
kube-system           kube-scheduler-rke2-pool1-nkjbq                                   1/1     Running     0          77m
kube-system           kube-scheduler-rke2-pool1-qmtm2                                   1/1     Running     0          84m
kube-system           rke2-coredns-rke2-coredns-65d668ddf9-g89rh                        1/1     Running     0          83m
kube-system           rke2-ingress-nginx-controller-8gwkl                               1/1     Running     0          80m
kube-system           rke2-ingress-nginx-controller-pxhqz                               1/1     Running     0          76m
kube-system           rke2-ingress-nginx-controller-vfq6s                               1/1     Running     0          75m
kube-system           rke2-metrics-server-6647ffc866-89wx4                              1/1     Running     0          81m
tigera-operator       tigera-operator-db8446457-btmqs                                   1/1     Running     0          83m
```
I'm joining a support bundle file:
[harvester-supportbundle_54a7b4d6-cc17-44c6-9fb4-c7d5f2ed53bd_2021-07-20T12-25-55Z.zip](https://github.com/harvester/harvester/files/6848363/harvester-supportbundle_54a7b4d6-cc17-44c6-9fb4-c7d5f2ed53bd_2021-07-20T12-25-55Z.zip)

**Comment by juadk** [comment]
The RKE1 issue #1030 is solved, I misconfigured the ssh user in the node template.
I filled the user of my ssh key which was created in Harvester instead of giving the default ssh user for the ubuntu image.
![126472555-0a5299b8-5af5-41bd-98a8-f3e7b0d5b99f](https://user-images.githubusercontent.com/6025636/126485884-969a0a4f-2d97-4fb3-ac0c-1ad45247805f.png)
I was tricked because VM, network and namespace are imported from Harvester, so I though the same for the ssh user.

**Comment by juadk** [comment]
I'm done with the deployment of RKE1 and RKE2 with the harvester node driver. As discussed with @guangbochen, I didn't check all the possibilities or features bring in the node template and so on.
I only made sure that the node driver is able to deploy RKE1 / RKE2 on Harvester.

**Overall status:**
![rke1-ok1](https://user-images.githubusercontent.com/6025636/126756369-00a13be6-2672-4933-be86-9227cb7a88c0.png)

**RKE1 cluster details:**
![rke1-ok2](https://user-images.githubusercontent.com/6025636/126756512-480bd187-0c20-4575-a049-b3970d65d96c.png)

**RKE2 cluster details:**
![rke2-ok1](https://user-images.githubusercontent.com/6025636/126756466-8f609c56-900d-4755-8cc2-82014185c96b.png)
